llvm 6: init

dtzWill opened the pull request:
###### Motivation for this change

Start building and testing LLVM 6, currently on RC1.

Helps ensure release doesn't have bugs re:NixOS / our projects, and hopefully enables us to at least make LLVM 6 available after final release and update the default shortly after.

Anyway, test test please! :)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

dtzWill committed:
llvm6: copy llvm5

dtzWill committed:
llvm6: rc1, hashes, top-level

dtzWill committed:
lld: Add dep on libxml2

grahamc commented:
I introduced a failure in the evaluator today, sorry about that. Testing for a fix:

@GrahamcOfBorg eval

dtzWill commented:
@GrahamcOfBorg build clang_6 lld_6 lldb_6

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34457&attempt_id=none)

<details><summary>Partial log (click to expand)</summary><p>

```
Apply anyway? [n]
Skipping patch.
2 out of 2 hunks ignored -- saving rejects to file lib/Support/Unix/Memory.inc.rej
builder for '/nix/store/i92fqldk238rl07g26bkjqxvkv94fra8-llvm-6.0.0rc1.drv' failed with exit code 1
cannot build derivation '/nix/store/b5jrr1h8b46ygimw9lp4drcnxn87hw8h-clang-6.0.0rc1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/irx1p6ad17rd4lnjzipyxk3m1d8209fm-clang-manpages-6.0.0rc1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/y1a9xzcngy03br51sbprglql0i5nb8ly-lld-6.0.0rc1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/6gz68lw4cxn3hdn07rx5k7darl2xw61p-clang-wrapper-6.0.0rc1.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/8r11d5hv3rqxv17038nj2fwkhw9n8wqx-lldb-6.0.0rc1.drv': 2 dependencies couldn't be built
error: build of '/nix/store/6gz68lw4cxn3hdn07rx5k7darl2xw61p-clang-wrapper-6.0.0rc1.drv', '/nix/store/8r11d5hv3rqxv17038nj2fwkhw9n8wqx-lldb-6.0.0rc1.drv', '/nix/store/y1a9xzcngy03br51sbprglql0i5nb8ly-lld-6.0.0rc1.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34457&attempt_id=none)

<details><summary>Partial log (click to expand)</summary><p>

```
[ 19%] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/InstructionSimplify.cpp.o
[ 19%] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/Interval.cpp.o
[ 19%] Building CXX object lib/Analysis/CMakeFiles/LLVMAnalysis.dir/IntervalPartition.cpp.o
building of '/nix/store/6sxf163jcv8cam64iv64dj0pc5jhz0g3-llvm-6.0.0rc1.drv' timed out after 1800 seconds
cannot build derivation '/nix/store/pkkf8vz9gxjbd064w8jz3z5mhc5zx2y0-clang-6.0.0rc1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/66swviph1yyx87n8kp0wg43d2s105lyj-clang-manpages-6.0.0rc1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/x3i2pb5vn93bc3alvj5rg6vhavpi5l8g-lld-6.0.0rc1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/1dfaybl6hm8p12mqbids4lz6aqbivg42-clang-wrapper-6.0.0rc1.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/nkx8mn730kg13wx16vqf7wfp13javjwg-lldb-6.0.0rc1.drv': 2 dependencies couldn't be built
error: build of '/nix/store/1dfaybl6hm8p12mqbids4lz6aqbivg42-clang-wrapper-6.0.0rc1.drv', '/nix/store/nkx8mn730kg13wx16vqf7wfp13javjwg-lldb-6.0.0rc1.drv', '/nix/store/x3i2pb5vn93bc3alvj5rg6vhavpi5l8g-lld-6.0.0rc1.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34457&attempt_id=a94d88b3-201b-4eb2-b91e-5a0d67ff88b2)

<details><summary>Partial log (click to expand)</summary><p>

```
[ 40%] Building CXX object lib/Target/AMDGPU/CMakeFiles/LLVMAMDGPUCodeGen.dir/AMDGPULibCalls.cpp.o
[ 40%] Building CXX object lib/Target/AMDGPU/CMakeFiles/LLVMAMDGPUCodeGen.dir/AMDGPULibFunc.cpp.o
[ 40%] Building CXX object lib/Target/AMDGPU/CMakeFiles/LLVMAMDGPUCodeGen.dir/AMDGPULowerIntrinsics.cpp.o
building of ‘/nix/store/03ikphp5a22crk0m6vmy231ydwy8qpq1-llvm-6.0.0rc1.drv’ timed out after 3600 seconds
cannot build derivation ‘/nix/store/4w3r4s8s4yy93sd3745yy68xzx0n3h83-clang-6.0.0rc1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/77h0vadd54g207r22m5rjzqaapvsgbg4-clang-manpages-6.0.0rc1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/aaarl8abw7dz2lpj193wamv6s9hz16kg-lld-6.0.0rc1.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/4cigd5f184c80nr3sdq7fyzfh383xk7v-clang-wrapper-6.0.0rc1.drv’: 2 dependencies couldn't be built
cannot build derivation ‘/nix/store/3w8gb4v5bia2wlg4c037cimk1lmssiyk-lldb-6.0.0rc1.drv’: 2 dependencies couldn't be built
error: build of ‘/nix/store/3w8gb4v5bia2wlg4c037cimk1lmssiyk-lldb-6.0.0rc1.drv’, ‘/nix/store/4cigd5f184c80nr3sdq7fyzfh383xk7v-clang-wrapper-6.0.0rc1.drv’, ‘/nix/store/aaarl8abw7dz2lpj193wamv6s9hz16kg-lld-6.0.0rc1.drv’ failed
```
</p></details>

LnL7 commented on pkgs/development/compilers/llvm/6/llvm.nix:
```diff
@@ -0,0 +1,171 @@
+    mv compiler-rt-* $sourceRoot/projects/compiler-rt
+  '';
+
+  outputs = [ "out" ]
```
Does this need something similar to https://github.com/NixOS/nixpkgs/pull/33871 to get python out of the closure?

dtzWill replied:
Good call.  Added.

LnL7 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6314,13 +6315,16 @@ with pkgs;
 
   llvm = llvmPackages.llvm;
 
+  llvm_6  = llvmPackages_6.llvm;
```
Should we keep exposing all of these top-level?

LnL7 commented on pkgs/development/compilers/llvm/6/llvm.nix:
```diff
@@ -0,0 +1,171 @@
+in stdenv.mkDerivation (rec {
+  name = "llvm-${version}";
+
+  unpackPhase = ''
```
I cleaned this up on a test branch recently, let me see if I can find it.

LnL7 reviewed (commented):
Everything builds on darwin.

dtzWill committed:
llvm6: no need to patch on aarch64, already applied

dtzWill committed:
llvm6: same python output fix from #33871

Unlike that PR, however, this doesn't need to go to staging
since no one depends on a package that doesn't exist yet :).

dtzWill commented:
rc2 has been tagged, but appears they don't upload the source tarballs until the "release testers" all build and upload binaries, so should be available in a few days.   Not worth bothering with grabbing off of SVN in the meantime, IMO.

LnL7 commented:
I don't think this is urgent since the build works.

dtzWill committed:
llvm6: rc1 -> rc2

dtzWill committed:
clang-6 test variants

dtzWill commented:
rebased and fixed merge conflicts.  Added testing variants like we have for other versions.

Will be adding musl compat fixes too, now that they're in master as well :+1:.

dtzWill commented:
Postponing musl fixes here while resolving discussion about musl ...

dtzWill committed:
llvm6: musl compat

dtzWill committed:
llvm6: drop perl and groff, as in llvm4/5

dtzWill committed:
llvm6: remove copied "remove this on next rebuild" value :)

dtzWill committed:
llvm6: rc2 -> rc3

dtzWill commented:
rebased, rc2 -> rc3 \o/

dtzWill commented:
Hold on to your lugnuts! LLVM 6 final has been tagged:

http://lists.llvm.org/pipermail/release-testers/2018-March/000667.html

Will be a few more days for them to prepare the release and official source tarballs but .... soon!!

LnL7 commented:
Great!

dtzWill committed:
llvm6: rc3 -> final

dtzWill commented:
Release notes: http://releases.llvm.org/6.0.0/docs/ReleaseNotes.html

Note this adds support for "retpolines", which may be of interest to those using clang-based stdenv's.

data-man commented:
Sorry, but how do I use the clang 6?
I just started using the NixOS.

dtzWill commented:
I don't think it's hit any channels yet, so in the meantime you can use something like this:

```console
$ NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs/archive/master.tar.gz nix-shell -p clang_6
```

and use `clang` within that shell.  Note that "installing" clang generally doesn't work as you may expect, there's an issue discussing this.
